support-models/src/main/scala/com/gu/support/workers/states/CreateZuoraSubscriptionState.scala

```diff
@@ -1,31 +1,100 @@
 package com.gu.support.workers.states
 
-import java.util.UUID
+import com.gu.i18n.Country
 
-import com.gu.salesforce.Salesforce.SalesforceContactRecords
+import java.util.UUID
+import com.gu.salesforce.Salesforce.{SalesforceContactRecords, SfContactId}
+import com.gu.support.encoding.Codec.deriveCodec
 import com.gu.support.promotions.PromoCode
 import com.gu.support.workers.{PaymentMethod, SalesforceContactRecord, User, _}
 import org.joda.time.LocalDate
-import com.gu.support.encoding.CustomCodecs.{decodeLocalTime, encodeLocalTime}
+import com.gu.support.encoding.CustomCodecs.{decodeLocalTime, encodeLocalTime, decodeCountry, encodeCountryAsAlpha2}
+import com.gu.support.encoding.{Codec, DiscriminatedType}
 import com.gu.support.redemptions.RedemptionData
+import com.gu.support.workers.GiftRecipient.{DigitalSubscriptionGiftRecipient, WeeklyGiftRecipient}
 
-case class CreateZuoraSubscriptionState(
+case class CreateZuoraSubscriptionWrapperState(
+  productSpecificState: CreateZuoraSubscriptionState,
   requestId: UUID,
   user: User,
-  giftRecipient: Option[GiftRecipient],
   product: ProductType,
   analyticsInfo: AnalyticsInfo,
-  paymentMethod: Either[PaymentMethod, RedemptionData],
   firstDeliveryDate: Option[LocalDate],
   promoCode: Option[PromoCode],
-  salesforceContacts: SalesforceContactRecords,
   acquisitionData: Option[AcquisitionData]
 ) extends FailureHandlerState
```

Comment:
(most) product specific state is removed and put into the productSpecificState field.

Some obvious ones are remaining and appear unused e.g. promo code and first devliery date.  This is because they are used by FailureHandlerState elasticsearch step as key-value pairs.  They are not needed as strongly typed data so can be moved into AnalyticsInfo with the other similar fields.  This can happen in another (reasonably small) PR rather than adding here.

Reply:
I notice that the product specific state does have a `product` field in it which duplicates the top level product, is it not straightforward to switch code over to using this?

Reply:
yes, everything is switched over within this lambda, the only reason it is still at the top level is because this state needs to deserialise directly as FailureHandlerState, as FailureHandler it needs to send the correct failure email to the user.
This won't be a problem to fix to use a suitable enum or even the productType value, but I think it makes more sense to do it separately.